src/GraphQL/Utilities/Visitors/SchemaValidationVisitor.cs

```diff
@@ -137,24 +149,32 @@ public override void VisitInputObject(IInputObjectGraphType type, ISchema schema
                 if (item.Count() > 1)
                     throw new InvalidOperationException($"The inpit field '{item.Key}' must have a unique name within Input Object type '{type.Name}'; no two fields may share the same name.");
             }
+        }
 
-            if (type.Fields?.Count > 0)
+        /// <inheritdoc/>
+        public override void VisitInputFieldDefinition(FieldType field, IInputObjectGraphType type, ISchema schema)
+        {
+            // 2.2
+            if (field.Name.StartsWith("__"))
+                throw new InvalidOperationException($"The input field '{field.Name}' of an Input Object '{type.Name}' must not have a name which begins with the __ (two underscores).");
+
+            // 2.3
+            if (field.ResolvedType != null ? field.ResolvedType.IsInputType() == false : field.Type?.IsInputType() == false)
+                throw new InvalidOperationException($"The input field '{field.Name}' of an Input Object '{type.Name}' must be an input type.");
+
+            // validate default value
+            if (field.DefaultValue != null && field.ResolvedType != null)
```

Comment:
Regarding `field.ResolvedType != null` check here and in `VisitFieldArgumentDefinition`. I think that no one field/argument after schema initialization may have `null` in `ResolvedType`. Let's add one more explicit check for `ResolvedType != null` and remove it from here and `VisitFieldArgumentDefinition`: ` if (field.DefaultValue != null ` / `if (argument.DefaultValue != null`

Reply:
#2338